reflect.go

```diff
@@ -0,0 +1,387 @@
+package main
+
+import (
+	"fmt"
+	"go/types"
+	"path/filepath"
+
+	"golang.org/x/exp/maps"
+	"golang.org/x/exp/slices"
+	"golang.org/x/tools/go/ssa"
+)
+
+var checkedAPIs = make(map[string]bool)
+
+// Record all instances of reflection use, and don't obfuscate types which are used in reflection.
+func (tf *transformer) recordReflection(ssaPkg *ssa.Package) {
+	if reflectSkipPkg[ssaPkg.Pkg.Name()] {
+		return
+	}
+
+	lenPrevKnownReflectAPIs := len(cachedOutput.KnownReflectAPIs)
+
+	// find all unchecked APIs to add them to checkedAPIs after the pass
+	notCheckedAPIs := make(map[string]bool)
+	for _, knownAPI := range maps.Keys(cachedOutput.KnownReflectAPIs) {
+		if !checkedAPIs[knownAPI] {
+			notCheckedAPIs[knownAPI] = true
+		}
+	}
+
+	tf.ignoreReflectedTypes(ssaPkg)
+
+	// all previously unchecked APIs have now been checked add them to checkedAPIs,
+	// to avoid checking them twice
+	maps.Copy(checkedAPIs, notCheckedAPIs)
+
+	// if a new reflectAPI is found we need to Re-evaluate all functions which might be using that API
+	if len(cachedOutput.KnownReflectAPIs) > lenPrevKnownReflectAPIs {
+		tf.recordReflection(ssaPkg)
+	}
+}
+
+func (tf *transformer) ignoreReflectedTypes(ssaPkg *ssa.Package) {
+	for _, memb := range ssaPkg.Members {
+		if t, ok := memb.(*ssa.Type); ok {
+			// methods aren't package members only their reciever types are
+			// so some logic is required to find the methods a type has
+
+			// yes, finding all methods really only works with both loops
```

Comment:
Hmm, that would be surprising to me, and at least misleading documentation in the go/ssa package.

https://go.dev/ref/spec#Method_sets says:

> The method set of a pointer to a defined type T (where T is neither a pointer nor an interface) is the set of all methods declared with receiver *T or T. 

That is, if you ask for the method set of `*T`, it should give you both kinds of methods at the same time, avoiding the need for two calls.

I think we should raise a bug upstream (ideally with a small reproducer) and link it here as a TODO.

Reply:
Tested it again right now still the case